### DISCUSSION TITLE
Customizing Linux config - dunfell-l4t-r32.5.0 (best pratice?)

### DISCUSSION THREAD

**Discussion by bkirkTag** [OE4T/meta-tegra#667] [Q&A] [answered]
Hi,
I'm having trouble customizing the linux configuration.
What is the best practice on changing the linux configuration?

My first try was adding a linux-tegra_%.bbappend, but this kconfig fragment isn't changing the linux defconfig:
```
FILESEXTRAPATHS_prepend := "${THISDIR}/${PN}:"
SRC_URI += "file://xxxxxxx.cfg" 
```

_xxxxxxxx.cfg_:
```
CONFIG_MD=y
CONFIG_BLK_DEV_DM=y
CONFIG_BLK_DEV_LOOP=y
CONFIG_DM_VERITY=y
CONFIG_SQUASHFS=y
```

If I replace "xxxxxxxx.cfg" with a full "defconfig" it looks like the resulting target configuration is correct but i don't rely on that because of WARNINGS during bitbake:
Can anyone tell me how to enable the extra CONFIGurations correct???

```
Build Configuration:
BB_VERSION           = "1.46.0"
BUILD_SYS            = "x86_64-linux"
NATIVELSBSTRING      = "universal"
TARGET_SYS           = "aarch64-poky-linux"
MACHINE              = "jetson-nano-devkit"
DISTRO               = "poky"
DISTRO_VERSION       = "3.1.6"
TUNE_FEATURES        = "aarch64 armv8a crc"
TARGET_FPU           = ""
meta                 
meta-poky            
meta-yocto-bsp       = "dunfell:d20ef1f5a5de7820b3e9b7f539b51f94954e9cd5"
meta-rauc            = "dunfell:100e8852fe9001215f649122d70a62a3fb47a50d"
meta-tegra           
contrib              = "dunfell-l4t-r32.5.0:ccf81f625992857ca04e7afcb2b1b3173262b9f6"
meta-qt5             = "dunfell:b4d24d70aca75791902df5cd59a4f4a54aa4a125"
meta-oe              
meta-python          
                     = "dunfell:346681e7bf9c78008a845fc89031be4fd4ceb3a1"
meta-custommm         = "dunfell_testing_rauc2:04c40eb10570f47b64114cb52ca0e15236353452"

Initialising tasks: 100% |##################################################################################################################################################################| Time: 0:00:04
Sstate summary: Wanted 22 Found 0 Missed 22 Current 3616 (0% match, 99% complete)
NOTE: Executing Tasks
WARNING: linux-tegra-4.9.201+gitAUTOINC+c68230c5b4-r0 do_kernel_metadata: defconfig detected in WORKDIR. tegra_defconfig skipped
WARNING: linux-tegra-4.9.201+gitAUTOINC+c68230c5b4-r0 do_kernel_metadata: [NOTE]: defconfig was supplied both via KBUILD_DEFCONFIG and SRC_URI. Dropping SRC_URI defconfig
NOTE: Tasks Summary: Attempted 8789 tasks of which 8478 didn't need to be rerun and all succeeded.

Summary: There were 2 WARNING messages shown.

```

**Comment by madisongh** [comment, marked as answer]
Your approach of adding a bbappend with a reference to your added config fragment file should work as described in the [Yocto Kernel Development Manual](https://docs.yoctoproject.org/kernel-dev/common.html#changing-the-configuration).  It won't alter the defconfig file, but is applied on top of the defconfig to produce the `.config` file used during the build.

If you really want to provide a full defconfig file in SRC_URI, you should also add `KBUILD_DEFCONFIG = ""` in your bbappend, since the `linux-tegra` recipe sets that variable by default.

**Reply by bkirkTag** [reply]
Thanks for a very fast reply. 
I will try the KBUILD_DEFCONFIG approach first :)

**Reply by madisongh** [reply]
OK, although I do recommend using config fragments to make it easier to migrate to new versions of the L4T BSP.  An example that I've used is [here](https://github.com/madisongh/tegra-test-distro/blob/master/layers/meta-testdistro/recipes-kernel/linux/).

**Reply by ichergui** [reply]
Agree with @madisongh 
Please follow the recommendation from [Yocto](https://docs.yoctoproject.org/kernel-dev/common.html#changing-the-configuration), and let us know if it works for you.

**Reply by bkirkTag** [reply]
You are right - the KBUILD_DEFCONFIG approach was just to check that my build warnings will be removed (and yes it fixes it).
Now I will go the for the 'fragment-way' by appending a .cfg file.

**Reply by bkirkTag** [reply]
My fault!!!

I mixed around with linux-tegra_%.bbappend  and linux-yocto_%.bbappend,  so my initial attempt to "patch" the config was on the wrong linux (linux-yocto    skipped).

The additional CONFIG_xxxx settings from my .cfg file now passes to the final .config file.

Thanks for the guidance!
